File: lesson_5_1/app/src/main/java/com/kalashnyk/denys/simplemvpapp/presentation/activity/main/MainActivity.java
package com.kalashnyk.denys.simplemvpapp.presentation.activity.main;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;

import com.kalashnyk.denys.simplemvpapp.R;
import com.kalashnyk.denys.simplemvpapp.data.database.entity.UserEntity;
import com.kalashnyk.denys.simplemvpapp.di.component.PresenterComponent;
import com.kalashnyk.denys.simplemvpapp.presentation.activity.detail.DetailActivity;
import com.kalashnyk.denys.simplemvpapp.presentation.adapter.UserAdapter;
import com.kalashnyk.denys.simplemvpapp.presentation.base.BaseActivity;
import com.kalashnyk.denys.simplemvpapp.presentation.item.IUserItemClickListener;

import java.util.List;

import javax.inject.Inject;

public class MainActivity extends BaseActivity implements MainContract.View {

    @Inject
    protected MainContract.Presenter mPresenter;

    private UserAdapter mAdapter;

    private IUserItemClickListener<UserEntity> mItemClickListener = new IUserItemClickListener<UserEntity>() {
        @Override
        public void openDetail(UserEntity productEntity) {
            openItemDetail(productEntity.getId());
        }
    };

    public static Intent newInstance(Context context) {
        Intent intent = new Intent(context, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK | Intent.FLAG_ACTIVITY_CLEAR_TASK);
        return intent;
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        initializeToolbar("Users", false);
        mPresenter.attachWithView(this);
        mPresenter.getUsers();
    }

    @Override
    protected void injectDependency(PresenterComponent presentersComponent) {
        presentersComponent.inject(this);
    }

    private void openItemDetail(long id) {
        getContext().startActivity(DetailActivity.newInstance(this, (int) id));
    }

    @Override
    public void showUsers(List<UserEntity> users) {
        RecyclerView rvUsers = findViewById(R.id.rv_users);
        LinearLayoutManager manager = new LinearLayoutManager(this);
        mAdapter = new UserAdapter(this, users);
        mAdapter.setItemClickListener(mItemClickListener);
        rvUsers.setLayoutManager(manager);
        rvUsers.setAdapter(mAdapter);

    }

}

File: lesson_9/app/src/main/java/com/kalashnyk/denys/kotlinbindingsample/presentation/activities/main/MainActivity.kt
package com.kalashnyk.denys.kotlinbindingsample.presentation.activities.main

import android.arch.lifecycle.Observer
import android.content.Context
import android.content.Intent
import android.databinding.DataBindingUtil
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import com.kalashnyk.denys.kotlinbindingsample.R
import com.kalashnyk.denys.kotlinbindingsample.databinding.ActivityMainBinding
import com.kalashnyk.denys.kotlinbindingsample.di.component.ViewModelComponent
import com.kalashnyk.denys.kotlinbindingsample.domain.AllUsersViewModel
import com.kalashnyk.denys.kotlinbindingsample.presentation.activities.detail.DetailActivity
import com.kalashnyk.denys.kotlinbindingsample.presentation.adapter.UserAdapter
import com.kalashnyk.denys.kotlinbindingsample.presentation.base.BaseActivity
import com.kalashnyk.denys.kotlinbindingsample.presentation.item.IUserItemClickListener
import com.kalashnyk.denys.kotlinbindingsample.repository.database.entity.UserEntity
import kotlinx.android.synthetic.main.activity_main.*
import javax.inject.Inject

class MainActivity : BaseActivity() {

    var viewModel: AllUsersViewModel? = null
        @Inject set

    lateinit var binding: ActivityMainBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        viewModel?.getAllItems()
        viewModel?.getLiveDataItems()?.observe(this, Observer { it?.let { initRecyclerView(it) } } )
    }

    private fun initRecyclerView(users: List<UserEntity>) {
        val manager = LinearLayoutManager(this)
        val userAdapter = UserAdapter(this, users, itemClickListener)
        userAdapter.setItemClickListener(itemClickListener)
        rvUsers.layoutManager = manager
        rvUsers.adapter = userAdapter
    }

    companion object {
        @JvmStatic
        fun newInstance(context: Context): Intent {
            val intent = Intent(context, MainActivity::class.java)
            intent.flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_CLEAR_TASK
            return intent
        }
    }

    private val itemClickListener = object : IUserItemClickListener<UserEntity> {
        override fun openDetail(entity: UserEntity) {
            openItemDetail(entity.id)
        }
    }

    override fun injectDependency(component: ViewModelComponent) {
        component.inject(this)
    }

    private fun openItemDetail(id: Int) {
        this.startActivity(DetailActivity.newInstance(this, id))
    }
}
